### DISCUSSION TITLE
Having trouble understanding code snippets with useTransition

### DISCUSSION THREAD

**Discussion by Szplugz** [pmndrs/react-spring#1943] [Support]
Hi, I'm trying to use react spring to make a simple photo carousel similar to [this](https://samuelbradley.com/portfolio/portfolio/) one. I basically want the images to fade in and out periodically, but I also want the ability to toggle between them manually with clicks. This is a long post, because I also want to understand the way some of these elements work.

There are 2 examples on the useTransition [page](https://react-spring.dev/hooks/use-transition#demos) that I think I can combine/customize to suit my needs. The first one is the standard image [fade](https://codesandbox.io/s/github/pmndrs/react-spring/tree/master/demo/src/sandboxes/image-fade?file=/src/App.tsx) example, and the other one is a toggle [animation](https://codesandbox.io/s/github/pmndrs/react-spring/tree/master/demo/src/sandboxes/simple-transition?file=/src/App.tsx). I think if I change the translate values in the toggle animation snippet, it will essentially create a unidirectional fade carousel.

Now, I'm having some trouble understanding the different parts of the code used in these snippets, because the documentation doesn't really clarify it for me. 

1. The docs say that the `transition` function accepts a callback that takes 4 arguments, but in both of those snippets, I see only two. Why is that? 

```javascript
transitions( ( style, i ) => {
    <animated.div
        className={styles.bg}
        style={{
          ...style,
          backgroundImage: `url(https://images.unsplash.com/${slides[i]}?w=1920&q=80&auto=format&fit=crop)`,
        }}
    />
))}
```

2. What exactly is the `style` parameter here? Is it the second parameter that's being passed to `useTransition()`? I.e
```javascript
 {  
    key: index,
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
    config: config.molasses,
    onRest: (_a, _b, item) => {
      if (index === item) {
        set(state => (state + 1) % slides.length)
      }
    },
    exitBeforeEnter: false,
}
```
If so, why does the spread operator need to be used? Furthermore, why does [this](https://react-spring.dev/basics#do-not-think-of-the-values-you-pass-as-styles-per-se) page say that these values are not "styles"? What does that mean?

3. In the toggle animation snippet, this is the body of the callback that `transitions()` takes in:

```javascript
const Page = pages[i]
return <Page style={style} />
```

Here, the `Page` element (let's take `pages[0]` for example) just translates to

```javascript
({ style }) => <animated.div style={{ ...style, background: 'lightpink' }}>A</animated.div>
```

so why is there a `style` attribute inside the `Page` component if the `animated.div` inside it already contains a `style` attribute?

I'm relatively new to React in general, so detailed answers would be greatly appreciated. Please assume I know nothing :)

_P.S There might be other libraries out there that are better suited for a simple photo carousel like this, but I plan to use react spring for other animations in the future, so I thought I'd start using it now._

**Comment by joshuaellis** [comment]
> The docs say that the transition function accepts a callback that takes 4 arguments, but in both of those snippets, I see only two. Why is that?

The transition function isn't well documented on the current docs. Check this out instead https://beta.react-spring.dev/docs/components/use-transition for (what i think) is a better explanation, happy to hear your thoughts.
The short answer though is this is the signature of the render function:

```ts
type TransitionFn = (
  style: SpringValues,
  item: Item,
  transitionState: TransitionState<Item>,
  index: number
) => ReactNode
```

> What exactly is the style parameter here? Is it the second parameter that's being passed to useTransition()? I.e

The style parameter is inferred on the props you passed depending on the current state of the transition e.g. `ENTERING` means we look at the `enter` property of the config object, again, see [this](https://beta.react-spring.dev/docs/components/use-transition#transition-function)

> Furthermore, why does [this](https://react-spring.dev/basics#do-not-think-of-the-values-you-pass-as-styles-per-se) page say that these values are not "styles"? What does that mean?

It makes the most sense for people to use `CSSProperties` as the keys to their springs because they're most familiar with them. But for instance you could call the key "apples" and interpolate that in several places, it's a pretty advanced concept of using `react-spring` imo, but helpful for lots of things.

> In the toggle animation snippet, this is the body of the callback that transitions() takes in:

I'm afraid reading your whole question, I don't understand what you're asking to help out on that one.
